### PR TITLE
WebGL rendering

### PR DESCRIPTION
**Pull request by swederik**
**Don't merge this, it is nowhere near complete. I'm just putting it here so we have somewhere to discuss it**

This builds off of @jpambrun's great work here: https://github.com/jpambrun/webgl-demo

**Live demos:**
http://rawgit.com/chafey/cornerstone/webgl/example/WebGL/index.html
http://rawgit.com/chafey/cornerstone/webgl/example/webGLtest/index.html

I have been playing with @jpambrun's demo webGL code and it does seem "easy" to get it up and working with cornerstone. The demo above already allows WWWC behaviour with a test image.

Some challenges we will have to deal with:
- Many parts of cornerstone use HTML5 canvas-specific calls, and they will have to be changed. One example was that renderGreyscale previously used an off-screen canvas to render to, whereas with WebGL we can now render directly to enabledElement.canvas
- Packing 16 bit integer data was cleverly done into two of the RGBA channels for the WebGL texture. This is then decoded in the fragment shader. We will need to write other shaders to handle the other cases (to start with, uint8 data for the exampleImageLoader, so I can stop requesting the sample data), such as RGB data

**Changes from @jpambrun's demo so far:**
- Vertex and fragment shaders are now stored as strings in Javascript
- Image slope / intercept, viewport windowCenter and windowWidth are now taken into account

**Changes (Sept 1st)**
- Moved relevant rendering code to specific WebGL folder
- Moved storedPixelData texture-generation functions to cornerstone.webgl.dataUtilities
- Added flag to prevent re-initializing WebGl
- Added an option to specify renderer in cornerstone.enable
- Changed guess datatype to make unsigned int use signed int shader
- Add touch drag for demo images
- Created an example with WebGL and Canvas version next to each other (see below)

**Changes (Sept 18th)**
- Improved fallback to canvas if WebGL is not available / has poor performance on the device
- WebGL initialization occurs on first element enable (with webgl option)
- Added handlers for WebGL Context Lost and Context Restored (wipes texture cache, re-initializes everything)

**Changes (Sept 21st)**
- Added WebGL Texture Cache with settable maximum size, similar to cornerstone's regular image cache mechanism. (need to figure out how to determine max VRAM to use on mobile devices to prevent crashes, if possible?)

**Changes (Sept 23rd)**
- Updated fragment shader for Int16 data. Negative values weren't being passed to the shader properly leading to bright values in the wrong location.
- Added a minimum window width of 1 to the int8 and int16 fragment shaders. Not sure this is the best solution, though.

**Side-by-side demo:**
http://rawgit.com/chafey/cornerstone/webgl/example/WebGL/index.html

**Known issues:**
- Setting window width between 0 and 1 appears to invert the image. I guess we have an issue in the grayscale fragment shaders. **Update (Sept 23rd): Fixed**
- Some Android devices appear to display Int16 and UInt16 images incorrectly (uint8 and RGB work fine). Attached an image from a Samsung tablet:
  screenshot_2015-10-05-21-00-01

### CODE REVIEW

**Comment by NicolasRannou** [comment]
Seems interesting.

I am currently putting together a library on top of ThreeJS, using dicomParser to handle the dicom parsing. Main motivation is that any kind of image processing is too slow in JavaScript without WebGL. Also, we want to work in 3D space in order to add cool features such as Volume Rendering and render 3D models.

(not ready for showtime by any mean but you can try to upload dcm images there as a test http://fnndsc.github.io/vjs/#hacks_ddrop)

If you go for WebGL, you might consider using an already existing library.

Anyway, I wanted to mention that we had a discussion with @pieper about writing re-usable GLSL shaders and one idea was to create a commonGLSL organization/directory which would contain all kind of shaders. People could then use it in Javascript, Python, C++, etc.

Best,

**Comment by swederik** [comment]
Nice, thanks for the link. It looks good so far. Yeah ThreeJS is great, makes sense to build something xtk-like on top of it. Sharing shaders makes sense too.

**Comment by jpambrun** [comment]
@swederik, I order to avoid the "click Load unsafe scripts" issue, you could just link witout https. e.g.:
http://rawgit.com/swederik/cornerstone/webgl/example/webGLtest/index.html

@NicolasRannou, thanks for sharing, your demo looks amazing!

I also looked at Three.js before implementing my crude demo. However, I though it was a very large dependency to add to cornerstone. I think I would prefer a simple 2d solution with few external dependencies. 

I noticed that your are also using GL shaders and packing 16bit into two 8bit channels. Three.js can fallback to 2D canvas when WebGL contexts are unavailable. Is this supported by your implementation ? If so, how?

Again, outstanding work @swederik and @NicolasRannou.

**Comment by fernandojsg** [comment]
Hi everyone!
I've developed a (another) DICOM viewer using webgl (http://fernandojsg.com/project/medikka-dicom-viewer/) and I find quite interesting the approach to include this on cornerstone which I think is a great project I wish I could had when I started my viewer :)
Regarding the use of webgl it benefits are obvious: its outstanding perfomance for 2D image processing. It's trivial to create a huge amount of filters, and WC/WW operations are just realtime without any pain even in big images.
But it has also several problems: It doesn't support (yet) shared resources, what means that you can't share textures created on different div instances (cornerstone elements), so if you want to paint the image in another canvas you "should" generate the webgl texture again.
Have you thought about it? Or does your approach just generate the image as long as you need to render it?
Other problem could be if you try to cache all the textures images, you're ending getting a lot of graphics card memory, so you need to have a duplicated cache in your viewer, one for the images instances itself (dicom raw, and metadata) and another one for the GL textures.

Regarding the use of external libraries, I've used and contributed to three.js and I know it quite good, and I'm not sure how useful could it be for a project like cornerstone. three.js is a huge library focused mainly on 3D with webgl (the canvas support has been deprecated), while it can paint 2D, it maybe include so much overhead in the way they do it for so simple task like a dicom viewer needs.
Basically you just need to paint a 2D quad with a texture attached to it, and some pixel shaders applied to it the devil comes in the resources management and here three.js doesn't help.
Maybe could be better to use a simple wrapper like http://twgljs.org/ that hides complexity from webgl while keeping it simple and without needing you to learn another framework or include its overhead.
three.js could be useful for 3D rendering once you get the geometry, maybe from the server in stl format or whatever, but I think this is a huge step forward that is not meant to be the focus of this library.

So If you need help with the webgl implementation I'm happy to help you with it!

Sorry for the long post hehe :yum:

**Comment by NicolasRannou** [comment]
Yes I played a bit around with packing 16 bits into the 8 bits RGBA texture. However that does support the fallback to 2D texture automatically. (Does't work with custom shaders)

I agree ThreeJS might not be useful for this project but if you decide to go down the WebGL way, having re-usable/sharable shaders would be great.

Another thing you have to keep in mind is that you will also have to port the "widgets" (to measure a distance, an angle, etc) to WebGL.
Ultimately you will have to re-write all the visualization part of cornerstone, then maintain both WebGL and Canvas2D up to date...

**Comment by fernandojsg** [comment]
@NicolasRannou I don't see the point on migrating widgets to webgl I would go for a much simple approach: have two canvas, one for 2D widgets and another just for the rendered image, so you just need to care in maintain two renderers for canvas and webgl. WebGL could be great for dealing with 3d and fast image processing but on painting 2D stuff, like lines, circles, dots, text, it's a pain. So in that way people could just keep creating widgets as right now.

**Comment by NicolasRannou** [comment]
@fernandojsg yes you are correct no need to re-do anything on the widgets since it will just be a "2D renderer" implemented in WebGL.

**Comment by swederik** [comment]
OK so it sounds like everyone is in agreement that two canvases is a good idea. We came to the same conclusion on the cornerstone mailing list (https://groups.google.com/forum/#!topic/cornerstone-platform/USyw9b-r0yU). I'll see if I can get that working in cornerstoneTools.

As for shared resources, no it's not something I've thought of at all. Displaying the exact same image in multiple elements isn't all that common anyway, so I am happy to ignore the issue for now. 

Glad to see we have some WebGL experts here to help with this push forward.

**Comment by fernandojsg** [comment]
@swederik maybe user won't show the same image on different viewports, but he can arrange to move the images from one viewport to another or change the layout, etc, so it could be a pain to regenerate the textures on those cases, also thinking in future features you may need to use textures on other viewports like for example load a PET/TAC in two viewports, and in a third generate a fusion between those two.
Btw I'm quite new with this project so I don't know how you use to discuss this issues, here on github? or in google groups?

**Comment by chafey** [comment]
For those interested, we will be doing a mini hackathon this Saturday to move this initiative forward!

https://groups.google.com/forum/#!topic/cornerstone-platform/hd-Aoms5SFg

**Comment by swederik** [comment]
Quick update since the hackathon:
- Int16 data is now stored in the WebGL texture alpha / luminance channels, instead of red/green from RGBA
- uint8 data is stored in a WebGL alpha-only texture
- RGB Images are stored in gl.RGB three-channel texture
- WebGL is drawn to an offscreen canvas and pixels are copied to the onscreen canvas using the HTML canvas drawImage method. This means that we don't need to change anything in cornerstoneTools. The demo link in the first post now works with pan / zoom, and displays a 4 megapixel image with sub-millisecond render times.

**Progress:**
- uint8 and int16 work fine, RGB works but windowing is broken
- Switching images breaks everything. We may need to release the canvas context or something?
- We need a mechanism for getting datatype info from the loaded image, so we know which renderer to use. For now I've hardcoded ````image.datatype = "uint8"```` in the example image loader.

We are well on our way to having WebGL support in cornerstone!

**Comment by pieper** [comment]
Awesome!

**Comment by fernandojsg** [comment]
Maybe you could merge it into `webgl`branch instead of `master`so to keep the branch updated but separated from master as this features will be unestable for a while.

**Comment by jpambrun** [comment]
This is awesome. Sorry I couldn't make it.

**Comment by swederik** [comment]
@fernandojsg Yep just sent a PR https://github.com/chafey/cornerstone/pull/27

**Comment by fernandojsg** [comment]
:+1:

**Comment by chafey** [comment]
Hi All - this is looking really good and I am hoping to merge it in soon.  A few thoughts:

Cornerstone is designed such that the image loader can control how an image is rendered.  Most image loaders will use one of the three renderers included in cornerstone core (renderColorImage, renderGrayscaleImage or renderWebImage).  I think the right thing for the WebGL path is to be subordinate to the existing renderGrayscaleImage and renderColorImage paths.  Currently the WebGL path is invoked at a higher level than this - from the drawImage level which gives precendence to the WebGL path over whatever the image loader may desire (which breaks the design).  So I would like to see the call refactored down a level

I also think we should make the second parameter to enable() an options object which can turn on WebGL mode, rather than just being a string.  This way we can pass in other advanced options that we may run into in the future.

@swederik mentioned that browsers put limitations on how many enabled elements can have WebGL enabled.  I am not exactly sure how to handle this one other than perhaps through documentation.  Gracefully falling back to canvas mode may not be the right thing - we should push such failure to the developer to handle.  Users don't like inconsistent performance so we don't want to be "too smart" about it.

**Comment by swederik** [comment]
Awesome. Your comments make sense. I'll try to sort these issues out ASAP.

How do you suggest pushing it back to the developer? An event like CornerstoneWebGLError or something?

**Comment by chafey** [comment]
Is it a resource allocation issue?  If so - can we allocate those resources in the enable() call and throw an exception if they cannot be allocated?  Hopefully we don't have to deal with errors that happen after enable() during a render call.

**Comment by swederik** [comment]
I haven't managed to find good documentation on it yet ("WARNING: Too many active WebGL contexts. Oldest context will be lost." in Chrome). I'll keep looking. 

For now I have done the refactoring as you suggested. It's now just an if statement inside renderGreyscaleImage and renderColorImage that determines where to retrieve the renderCanvas from.

I also made a branch of OHIF/Viewers with this version of the WebGL renderer (https://github.com/OHIF/Viewers/tree/cornerstoneWebGL), in case you want to test it with a real application (e.g. with 9 viewports).

 Let me know what you think.
